File: framework/java/src/main/java/soda/unittest/job/codec/CodecFactory.java
package soda.unittest.job.codec;

import java.util.HashMap;
import java.util.Map;

public class CodecFactory {
	
	private static Map<Class<?>, Class<? extends ICodec<?,?>>> codecMap;
	
	static {
		codecMap = new HashMap<>();
		codecMap.put(int[].class, IntArrayCodec.class);
		codecMap.put(int[][].class, IntArray2dCodec.class);
        codecMap.put(char.class, CharCodec.class);
        codecMap.put(Character.class, CharCodec.class);
	}

	public static ICodec<?,?> create(Class<?> objClass) throws Exception {
		Class<? extends ICodec<?,?>> cls = codecMap.get(objClass);
		if (cls != null) {
			return cls.getDeclaredConstructor().newInstance();
		}
		
		if (objClass.isArray()) {
			return new ObjectArrayCodec(getElementType(objClass), getDimension(objClass));
		}
		
		return new DefaultCodec();
	}
	
	private static Class<?> getElementType(Class<?> cls) {
		return cls.isArray() ? getElementType(cls.getComponentType()) : cls;
	}
	
	private static int getDimension(Class<?> cls) {
		return cls.isArray() ? 1 + getDimension(cls.getComponentType()) : 0;
	}
	
}

File: framework/scala/src/main/scala/soda/scala/unittest/codec/CodecFactory.scala
package soda.scala.unittest.codec

import collection.mutable.{Map => MMap}

object CodecFactory {

    val codecMap = MMap[Class[_], Class[_ <: ICodec[_,_]]](
            classOf[Array[Int]] -> classOf[IntArrayCodec],
            classOf[Array[Array[Int]]] -> classOf[IntArray2dCodec]
    )

    def create(objClass: Class[_]): ICodec[_,_] = {
        val cls = codecMap.get(objClass)
        cls match {
            case Some(s) => s.getDeclaredConstructor().newInstance()
            case None => {
                if (objClass.isArray())
                    new ObjectArrayCodec(getElementType(objClass), getDimension(objClass))
                else
                    new DefaultCodec()
            }
        }
    }

    def getElementType(cls: Class[_]): Class[_] = {
        if (cls.isArray()) getElementType(cls.getComponentType()) else cls
    }

    def getDimension(cls: Class[_]): Int = {
        if (cls.isArray()) 1 + getDimension(cls.getComponentType()) else 0
    }

}
